### PR TITLE
Separate shouldDoReduction from doReduction for strings ext solver

### PR DESCRIPTION
**Pull request by ajreynol**
Towards model-based variant of core strings solver.  

This will require knowing if there is an extended function reduction waiting, in which case we should not try to build a model.  This reorganizes the extended functions string solver in preparation for this.

There are no intended behavior changes in this PR.

### CODE REVIEW

File: src/theory/strings/extf_solver.cpp [hunk 1]
```diff
@@ -80,10 +80,10 @@ ExtfSolver::ExtfSolver(Env& env,
 
 ExtfSolver::~ExtfSolver() {}
 
-bool ExtfSolver::doReduction(int effort, Node n)
+bool ExtfSolver::shouldDoReduction(int effort, Node n, int pol)
 {
-  Trace("strings-extf-debug")
-      << "doReduction " << n << ", effort " << effort << std::endl;
+  Trace("strings-extf-debug") << "shouldDoReduction " << n << ", pol " << pol
+                              << ", effort " << effort << std::endl;
```
Comment (yoni206) [inline src/theory/strings/extf_solver.cpp:86]: is this formatted?

File: src/theory/strings/extf_solver.cpp [hunk 2]
```diff
@@ -253,9 +284,19 @@ void ExtfSolver::checkExtfReductions(int effort)
     Trace("strings-extf-debug")
         << "  check " << n
         << ", active in model=" << d_extfInfoTmp[n].d_modelActive << std::endl;
-    bool ret = doReduction(effort, n);
-    if (ret)
+    // polarity, 1: positive, -1: negative, 0: neither
```
Comment (yoni206) [inline src/theory/strings/extf_solver.cpp:287]: I suggest moving this to the declaration in the h file.

**Review by yoni206 — APPROVED** [review]
LGTM, one minor comment.

File: src/theory/strings/extf_solver.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/strings/extf_solver.cpp:86]: Yes

**Pull request merged**